File: src/main/scala/com/ajjpj/amapper/core/impl/IdentityCache.scala
package com.ajjpj.amapper.core.impl

import java.util.IdentityHashMap
import scala.collection.convert.Wrappers
import com.ajjpj.amapper.core.{AMapperException, PathBuilder}

/**
 * @author arno
 */
class IdentityCache {
  private val map = Wrappers.JMapWrapper (new IdentityHashMap[AnyRef, AnyRef]())

  def register(source: AnyRef, target: AnyRef, path: PathBuilder) {
    map.get(source) match {
      case Some(prev) =>
        if(prev eq target)
          throw new AMapperException(s"duplicate registration of same target $target for $source", path.build)
        else
          throw new  AMapperException(s"duplicate registration of different target $target for $source (was: $prev)", path.build)
      case None => map += (source -> target)
    }
  }

  def get(source: AnyRef): Option[AnyRef] = map.get(source)
}

File: src/main/java/com/ajjpj/amapper/core/impl/IdentityCache.java
package com.ajjpj.amapper.core.impl;

import com.ajjpj.abase.collection.immutable.AOption;
import com.ajjpj.amapper.core.exclog.AMapperExceptionHandler;
import com.ajjpj.amapper.core.path.APath;
import com.ajjpj.amapper.core.tpe.AQualifiedType;

import java.util.HashMap;
import java.util.IdentityHashMap;
import java.util.Map;


/**
 * @author arno
 */
class IdentityCache {
    private final Map<IdentityKey, Object> map = new HashMap<> ();

    public void register(Object source, Object target, AQualifiedType targetType, APath path) {
        final IdentityKey key = new IdentityKey (source, targetType);

        if (map.containsKey (key)) {
            final Object prev = map.get(key);
            if(prev == target) {
                AMapperExceptionHandler.onError("duplicate registration of same target " + target + " for " + source, path);
            }
            else {
                AMapperExceptionHandler.onError("duplicate registration of different target " + target + " and " + prev + " for " + source, path);
            }
        }
        else {
            map.put (key, target);
        }
    }

    public AOption<Object> get(Object source, AQualifiedType targetType) {
        final IdentityKey key = new IdentityKey (source, targetType);

        if (map.containsKey (key)) {
            return AOption.some (map.get (key));
        }
        else {
            return AOption.none();
        }
    }

    private static class IdentityKey {
        final Object sourceObject;
        final AQualifiedType targetType;

        private IdentityKey (Object sourceObject, AQualifiedType targetType) {
            this.sourceObject = sourceObject;
            this.targetType = targetType;
        }

        @Override
        public boolean equals (Object o) {
            final IdentityKey that = (IdentityKey) o;
            return sourceObject == that.sourceObject && targetType.equals (that.targetType);
        }
        @Override
        public int hashCode () {
            int result = System.identityHashCode (sourceObject);
            result = 31 * result + targetType.hashCode ();
            return result;
        }
    }
}
